src/experimental-bots/filler/fillerMultithreaded.ts

```diff
@@ -1653,162 +1681,211 @@ export class FillerMultithreaded {
+									marketIds
+								)}: ${err.message}`
+							);
+							console.error(err);
 						}
-					} catch (err) {
-						if (!(err instanceof Error)) {
-							return;
-						}
-						const errorCode = getErrorCode(err) ?? 0;
-						logger.error(
-							`Error code: ${errorCode} while settling pnls for markets ${JSON.stringify(
-								marketIds
-							)}: ${err.message}`
-						);
-						console.error(err);
 					}
-				}
-				try {
-					const txs = await Promise.all(settlePnlPromises);
-					for (const tx of txs) {
-						logger.info(
-							`Settle positive PNLs tx: https://solscan/io/tx/${tx.txSig}`
-						);
+					try {
+						const txs = await Promise.all(settlePnlPromises);
+						for (const tx of txs) {
+							logger.info(
+								`Settle positive PNLs tx: https://solscan/io/tx/${tx.txSig}`
+							);
+						}
+					} catch (e) {
+						logger.error(`Error settling positive pnls: ${e}`);
 					}
-				} catch (e) {
-					logger.error(`Error settling positive pnls: ${e}`);
+					this.lastSettlePnl = now;
 				}
-				this.lastSettlePnl = now;
+			}
+
+			if (isTotalUnsettledPnlSettlable) {
+				await this.rebalance();
 			}
 		}
-		if (this.rebalanceFiller) {
-			logger.info(`Rebalancing filler`);
-			const fillerSolBalance = await this.driftClient.connection.getBalance(
-				this.driftClient.authority
-			);
+	}
 
-			this.hasEnoughSolToFill = fillerSolBalance >= this.minimumAmountToFill;
+	protected async rebalance() {
+		logger.info(`Rebalancing filler`);
+		const fillerSolBalance = await this.driftClient.connection.getBalance(
+			this.driftClient.authority
+		);
 
-			if (!this.hasEnoughSolToFill && this.jupiterClient !== undefined) {
-				logger.info(`Swapping USDC for SOL to rebalance filler`);
-				swapFillerHardEarnedUSDCForSOL(
-					this.priorityFeeSubscriber,
-					this.driftClient,
-					this.jupiterClient,
-					await this.getBlockhashForTx()
-				).then(async () => {
-					const fillerSolBalanceAfterSwap =
-						await this.driftClient.connection.getBalance(
-							this.driftClient.authority,
-							'processed'
-						);
-					this.hasEnoughSolToFill =
-						fillerSolBalanceAfterSwap >= this.minimumAmountToFill;
-				});
-			} else {
-				this.hasEnoughSolToFill = true;
-			}
+		this.hasEnoughSolToFill = fillerSolBalance >= this.minimumAmountToFill;
+
+		if (this.jupiterClient !== undefined) {
+			logger.info(`Swapping USDC for SOL to rebalance filler`);
+			swapFillerHardEarnedUSDCForSOL(
+				this.priorityFeeSubscriber,
+				this.driftClient,
+				this.jupiterClient,
+				await this.getBlockhashForTx()
+			).then(async () => {
+				const fillerSolBalanceAfterSwap =
+					await this.driftClient.connection.getBalance(
+						this.driftClient.authority,
+						'processed'
+					);
+				this.hasEnoughSolToFill =
+					fillerSolBalanceAfterSwap >= this.minimumAmountToFill;
+			});
+		} else {
+			this.hasEnoughSolToFill = true;
 		}
```

Comment:
we should throw if rebalane is toggled on but there's no jupiter client initialized